File: generating_mlir.py
from common import save_mlir  # caps GPU memory before importing jax; provides save_mlir / load_mlir

import jax
import jax.numpy as jnp


x = jnp.array([1.0, 2.0, 3.0], dtype=jnp.float32)
list(x.devices())

jax.jit(jnp.sum).lower(x).as_text()

# jaxpr: JAX's typed IR, one level above StableHLO
jax.make_jaxpr(jnp.sum)(x)


from userfuncs import negsumsq

jax.make_jaxpr(negsumsq)(x)           # jaxpr: integer_pow, reduce_sum, neg
jax.jit(negsumsq).lower(x).as_text()  # ... then lowered to StableHLO

save_mlir("negsumsq_jax.mlir", negsumsq, x)

File: generating_mlir.jl
include("common.jl")   # caps GPU memory, then defines save_mlir / load_mlir

using Enzyme: Enzyme
import ForwardDiff

x = Float32[1, 2, 3]
rx = to_rarray(x)

XLA.device(rx)

@code_hlo sum(rx)


include("userfuncs.jl")

negsumsq(x)

@code_llvm negsumsq(x)
save_llvm("negsumsq.ll", negsumsq, x)

ForwardDiff.gradient(negsumsq, x)

@code_hlo optimize=false negsumsq(rx)
@code_hlo negsumsq(rx)

save_mlir("negsumsq_reactant.mlir", negsumsq, rx)

loss = @jit negsumsq(rx)

@code_hlo Enzyme.gradient(Enzyme.Reverse, negsumsq, rx)

∇loss = @jit Enzyme.gradient(Enzyme.Reverse, negsumsq, rx)
